### ISSUE TITLE
Upgrade actions without expirations fail

### ISSUE THREAD

**Issue by criamico** [elastic/fleet-server#3103]
### Summary
[It was discovered](https://github.com/elastic/kibana/issues/170322#issuecomment-1810483659) that an UPGRADE action with no expiration doesn't get dispatched to elastic agent, as fleet-server dispatches only actions where `expiration > now`. [See logic here](https://github.com/elastic/fleet-server/blob/a465d093a882d8b4a9ec1ae97d8494b0d2f9ae39/internal/pkg/dl/actions.go#L117) 

So we should allow actions without expiration so in the future it will be possible to completely remove the expiration from Fleet.

**Comment by michel-laterman** [comment]
This should be a simple change on the fleet-server, but will need changes on the elastic-agent.

The assumption we made when initially implementing scheduled actions was that they need to expire in order to support a maintenance window.

The agent defines a scheduled action as one that has a start time and expiration, see [the interface](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/fleetapi/action.go#L56-L62).
Action's are queued in the dispatcher [here](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/dispatcher/dispatcher.go#L184) and only needs the `StartTime` function.
However when they are dequed the `Expiration` function [is called](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/dispatcher/dispatcher.go#L221).

It may be easiest for the elastic-agent to change the upgrade action's [Expiration function](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/fleetapi/action.go#L290) to return an empty/zero time value and nil error if Expiration is empty.

cc @cmacknz

**Comment by cmacknz** [comment]
> It may be easiest for the elastic-agent to change the upgrade action's [Expiration function](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/fleetapi/action.go#L290) to return an empty/zero time value and nil error if Expiration is empty.

Yes we need to make it so that actions that don't expire are considered valid. The unfortunate thing is that if we fix this agents without the fix will still consider actions without expirations as errors.

So we are probably better off setting the expiration time in Fleet to the longest possible expiration time. Setting an action to expire in 2099 or whatever is effectively the same as having it never expire.

**Comment by ycombinator** [comment]
Has this issue been resolved by https://github.com/elastic/kibana/pull/170879?

**Comment by jlind23** [comment]
@criamico any comment on https://github.com/elastic/fleet-server/issues/3103#issuecomment-1863372098 ?

**Comment by cmacknz** [comment]
> It may be easiest for the elastic-agent to change the upgrade action's [Expiration function](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/fleetapi/action.go#L290) to return an empty/zero time value and nil error if Expiration is empty.

We should eventually make this change in the agent, but I don't think it's worth changing Fleet given the backwards compatibility concern. We can't get rid of the expiration time until we know every agent version that expects it is no longer in use. For this same reason I don't think making any changes here is a priority.

I would leave this issue open as documentation about this problem. I'll edit the description to seem less impactful.
